src/providers/kubevirt/mod.rs

```diff
@@ -0,0 +1,271 @@
+        let metadata = self.read_metadata()?;
+        Self::known_attributes(metadata)
+    }
+
+    fn hostname(&self) -> Result<Option<String>> {
+        let metadata = self.read_metadata()?;
+        let hostname = if metadata.hostname.is_empty() {
+            None
+        } else {
+            Some(metadata.hostname)
+        };
+        Ok(hostname)
+    }
+
+    fn ssh_keys(&self) -> Result<Vec<PublicKey>> {
+        let metadata = self.read_metadata()?;
+        Self::public_keys(metadata)
+    }
+
+    fn networks(&self) -> Result<Vec<network::Interface>> {
+        warn!("network interfaces metadata requested, but not supported on this platform");
+        Ok(vec![])
+    }
+
+    fn virtual_network_devices(&self) -> Result<Vec<network::VirtualNetDev>> {
+        warn!("virtual network devices metadata requested, but not supported on this platform");
+        Ok(vec![])
+    }
+
+    fn boot_checkin(&self) -> Result<()> {
+        warn!("boot check-in requested, but not supported on this platform");
+        Ok(())
+    }
+}
+
+impl Drop for KubeVirtProvider {
+    fn drop(&mut self) {
+        if let Err(e) = crate::util::unmount(
+            self.temp_dir.path(),
+            3, // maximum retries
+        ) {
+            slog_scope::error!("failed to unmount kubevirt config-drive: {}", e);
+        };
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::Cursor;
+
+    #[test]
+    fn test_kubevirt_basic_attributes() {
+        let metadata = r#"
+{
+  "hostname": "test_instance-kubevirt.foo.cloud",
+  "uuid": "41b4fb82-ca29-11eb-b8bc-0242ac130003"
+}
+"#;
+
+        let bufrd = BufReader::new(Cursor::new(metadata));
+        let parsed = KubeVirtProvider::parse_metadata(bufrd).unwrap();
+        assert_eq!(parsed.instance_id, "41b4fb82-ca29-11eb-b8bc-0242ac130003");
+        assert_eq!(parsed.hostname, "test_instance-kubevirt.foo.cloud");
+
+        let attrs = KubeVirtProvider::known_attributes(parsed).unwrap();
+        assert_eq!(attrs.len(), 2);
+        assert_eq!(
+            attrs.get("KUBEVIRT_INSTANCE_ID"),
+            Some(&"41b4fb82-ca29-11eb-b8bc-0242ac130003".to_string())
+        );
+        assert_eq!(
+            attrs.get("KUBEVIRT_HOSTNAME"),
+            Some(&"test_instance-kubevirt.foo.cloud".to_string())
+        );
+    }
+
+    #[test]
+    fn test_kubevirt_extended_attributes() {
+        let metadata = r#"
+{
+  "hostname": "test_instance-kubevirt.foo.cloud",
+  "uuid": "41b4fb82-ca29-11eb-b8bc-0242ac130003",
+  "instance_type": "some_type"
+}
+"#;
+
+        let bufrd = BufReader::new(Cursor::new(metadata));
+        let parsed = KubeVirtProvider::parse_metadata(bufrd).unwrap();
+        assert_eq!(parsed.instance_id, "41b4fb82-ca29-11eb-b8bc-0242ac130003");
+        assert_eq!(parsed.hostname, "test_instance-kubevirt.foo.cloud");
+        assert_eq!(parsed.instance_type.as_deref().unwrap_or_default(), "some_type");
```

Comment:
Tiny nit: you can just use `unwrap()` here.  Failing tests are allowed to panic.  :slightly_smiling_face: